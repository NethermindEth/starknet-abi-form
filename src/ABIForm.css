#example-button {
  color: blue;
  font-size: 24px;
}
