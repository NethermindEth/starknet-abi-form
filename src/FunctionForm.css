.border-array-item {
  border: 1px solid limegreen;
  border-radius: 0.5rem;
  padding: 0.3rem 1rem;
  margin: 0.2rem 0rem;
}

.border-struct {
  border: 1px solid red;
  border-radius: 0.5rem;
  padding: 0.3rem 1rem;
  margin: 0.2rem 0rem;
}

.input-wrapper {
  margin: 0.5rem 0rem;
}

.btn {
  display: block;
  width: max-content;
  padding: 0.5rem;
  font-size: 1rem;
  border-radius: 0.5rem;
  border: 1px solid black;
}

.btn:hover {
  cursor: pointer;
  box-shadow: 1px 1px 5px black;
}

.input-error {
  color: red;
}
