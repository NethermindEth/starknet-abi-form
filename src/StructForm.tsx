// import React from 'react';
// import { ABIEnum, ABIStruct, yupAbiStructSchema } from './types';
// import { isACoreType } from './types/dataTypes';
// import {
//   extractSubTypesFromType,
//   hasArrayOfSubType,
//   hasSubTypes,
// } from './types/helper';

// type IStructForm = {
//   structParamName: string;
//   structAbi: ABIStruct;
//   structs?: ABIStruct[];
//   enums?: ABIEnum[];
// };

// const StructForm: React.FC<IStructForm> = ({
//   structParamName,
//   structAbi,
//   structs,
//   enums,
// }) => {
//   try {
//     yupAbiStructSchema.validateSync(structAbi);
//   } catch (e) {
//     return <p>Not a valid struct schema</p>;
//   }
//   console.log(structAbi);
//   return (
//     <div style={{ border: '1px solid red', padding: '0.5rem 0.1rem' }}>
//       <p>
//         {structParamName} | {structAbi?.name}
//       </p>
//       {structAbi?.members.map((structMember) => {
//         if (isACoreType(structMember?.type))
//           return (
//             <div
//               style={{
//                 display: 'flex',
//                 flexDirection: 'column',
//                 border: '1px solid lime',
//                 padding: '0.1rem',
//               }}
//             >
//               <label>
//                 {structMember?.name} | {structMember?.type}
//               </label>
//               <input placeholder={structMember?.name} />
//             </div>
//           );

//         const structIdx =
//           structs?.findIndex((struct) => struct.name === structMember?.type) ||
//           -1;

//         if (structIdx !== -1 && structs) {
//           return (
//             <StructForm
//               structParamName={structMember?.name}
//               structAbi={structs[structIdx]}
//               structs={structs}
//               enums={enums}
//             />
//           );
//         }

//         // Check if is a array
//         if (hasArrayOfSubType(structMember?.type)) {
//           const isSubTypes = extractSubTypesFromType(structMember?.type);
//           if (
//             isSubTypes?.contains &&
//             isSubTypes?.types &&
//             isSubTypes?.types?.length > 0
//           ) {
//             const arrSubType = isSubTypes?.types[0]; // Safe to take 1 element only, since we can only have homogenous arrays
//             console.log(structMember?.name);
//             console.log('Found Array of types: ', arrSubType);

//             if (isACoreType(arrSubType)) {
//               return (
//                 <div
//                   style={{
//                     display: 'flex',
//                     flexDirection: 'column',
//                     border: '1px solid lime',
//                     padding: '0.1rem',
//                   }}
//                 >
//                   <label>
//                     {structMember?.name} | {structMember?.type}
//                   </label>
//                   <input placeholder={`array: ${arrSubType}`} />
//                 </div>
//               );
//             }

//             const structIdx =
//               structs?.findIndex((struct) => struct.name === arrSubType) || -1;

//             if (structIdx !== -1 && structs) {
//               return (
//                 <StructForm
//                   structParamName={`array<${structMember?.name}>`}
//                   structAbi={structs[structIdx]}
//                   structs={structs}
//                   enums={enums}
//                 />
//               );
//             }
//           }
//         }

//         // Check if have subTypes
//         if (hasSubTypes(structMember?.type)) {
//           console.log('Found SubTypes', structMember.type);
//         }
//       })}
//     </div>
//   );
// };

// export default StructForm;
