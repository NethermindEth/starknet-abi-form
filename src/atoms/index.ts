import { atomWithStorage } from 'jotai/utils';

const ABI_FORM = 'STARKNET_ABI_FORM';

export const formsAtom = atomWithStorage<Record<string, {}>>(ABI_FORM, {});
