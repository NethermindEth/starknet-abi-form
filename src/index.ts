import './index.css';

export * from './types';

export * from './ABIForm';

export * from './ConstructorForm';
