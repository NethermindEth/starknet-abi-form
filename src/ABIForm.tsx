import './ABIForm.css';

import React, { useMemo, useState } from 'react';

import { ABI, abiSchema } from './types/index';
import {
  extractEnumsFromABI,
  extractFunctionFromRawAbi,
  extractStructFromABI,
  segregateViewAndExternalFunctions,
} from './types/helper';
import FunctionForm from './FunctionForm';
import { Content, List, Root, Trigger } from './UIComponents/Tabs/Tabs';
import { ActiveTabClasses, DefaultTabClasses } from './utils/tailwindClasses';
import { Provider } from './UIComponents/Tooltip/Tooltip';

export type CallbackReturnType = {
  functionName: string;
  raw: {};
  starkli?: {
    bigint: bigint[];
    decimal: string;
    hex: string;
  };
  starknetjs?: any[];
  stateMutability: 'view' | 'external';
};

export type ABIFormProps = {
  abi?: ABI;
  callBackFn: (value: CallbackReturnType) => void;
  responses?: Record<string, React.ReactNode>;
};

export const ABIForm: React.FC<ABIFormProps> = ({
  abi,
  callBackFn,
  responses,
}) => {
  try {
    abiSchema.validateSync(abi);
  } catch (e) {
    console.error(e);
    return <p className="invalid-abi">Not a Valid ABI Schema Cairo v2</p>;
  }

  // If abi validated successfully then it's a valid abi as per spec.
  // https://github.com/starkware-libs/starknet-specs/blob/master/api/starknet_metadata.json

  const { viewFunctions, externalFunctions } = useMemo(() => {
    try {
      const rawFns = extractFunctionFromRawAbi(abi, []);
      const filteredFunction = segregateViewAndExternalFunctions(rawFns);
      return filteredFunction;
    } catch (e) {
      return {
        viewFunctions: [],
        externalFunctions: [],
      };
    }
  }, [abi]);

  const structs = useMemo(() => {
    try {
      return extractStructFromABI(abi);
    } catch (e) {
      return [];
    }
  }, [abi]);

  const enums = useMemo(() => {
    try {
      return extractEnumsFromABI(abi);
    } catch (e) {
      return [];
    }
  }, [abi]);

  const [activeTab, setActiveTab] = useState<'read' | 'write'>('read');

  return (
    <Provider>
      <Root value={activeTab} className="tab-root">
        <List className="text-sm font-medium text-gray-500 border-b border-gray-200 dark:text-gray-400 dark:border-gray-700 tab-triggers-wrapper">
          <Trigger
            value="read"
            className={`${
              activeTab === 'read' ? ActiveTabClasses : DefaultTabClasses
            } tab-trigger`}
            onClick={() => {
              setActiveTab('read');
            }}
          >
            Read
          </Trigger>
          <Trigger
            value="write"
            className={`${
              activeTab === 'write' ? ActiveTabClasses : DefaultTabClasses
            } tab-trigger`}
            onClick={() => {
              setActiveTab('write');
            }}
          >
            Write
          </Trigger>
        </List>
        <Content value="read" className="tab-content">
          {viewFunctions.map((viewFn) => (
            <FunctionForm
              key={`viewFn-${viewFn.name}`}
              functionAbi={viewFn}
              structs={structs}
              callbackFn={callBackFn}
              response={responses && responses[viewFn?.name]}
              enums={enums}
            />
          ))}
        </Content>
        <Content value="write" className="tab-content">
          {externalFunctions.map((externalFn) => (
            <FunctionForm
              key={`externalFn-${externalFn.name}`}
              functionAbi={externalFn}
              structs={structs}
              callbackFn={callBackFn}
              response={responses && responses[externalFn?.name]}
              enums={enums}
            />
          ))}
        </Content>
      </Root>
    </Provider>
  );
};
