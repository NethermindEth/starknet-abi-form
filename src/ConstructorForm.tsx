import './ABIForm.css';

import React, { useMemo } from 'react';

import { ABI, abiSchema } from './types/index';
import {
  convertConstructorToFunction,
  EMPTY_CONSTRUCTOR_FUNCTION,
  extractConstructorFromRawAbi,
} from './types/helper';
import FunctionForm from './FunctionForm';
import { Provider } from './UIComponents/Tooltip/Tooltip';
import { CallbackReturnType } from './ABIForm';

export type ConstructorFormProps = {
  abi?: ABI;
  callBackFn: (value: CallbackReturnType) => void;
};

export const ConstructorForm: React.FC<ConstructorFormProps> = ({
  abi,
  callBackFn,
}) => {
  try {
    abiSchema.validateSync(abi);
  } catch (e) {
    console.error(e);
    return <p className="invalid-abi">Not a Valid ABI Schema Cairo v2</p>;
  }

  const constructor = useMemo(() => {
    try {
      const constructorFromAbi = extractConstructorFromRawAbi(abi);
      return convertConstructorToFunction(constructorFromAbi);
    } catch (e) {
      return EMPTY_CONSTRUCTOR_FUNCTION;
    }
  }, [abi]);

  return (
    <Provider>
      <FunctionForm
        callbackFn={callBackFn}
        functionAbi={constructor}
        structs={[]}
        buttonLabel="Deploy"
      />
    </Provider>
  );
};

export default ConstructorForm;
