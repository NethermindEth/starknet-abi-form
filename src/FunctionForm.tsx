import React, { useEffect, useState } from 'react';

import * as Yup from 'yup';
import { useFormik } from 'formik';
import loadashFp from 'lodash';
import { useAtom } from 'jotai';
import { ABIFunction, ABIStruct, yupAbiFunctionSchema } from './types';

import './FunctionForm.css';
import {
  extractAbiTypes,
  extractInitialValues,
  extractValidationSchema,
  reduceFunctionInputs,
} from './types/uiHelpers';
import {
  AccordionRoot,
  AccordionContent,
  AccordionItem,
  AccordionTrigger,
} from './UIComponents/Accordian/Accordian';
import { Button, ButtonColorsClasses } from './UIComponents/Button/Button';
import Tag, { TagColors } from './UIComponents/Tag/Tag';
import {
  flattenArrays,
  flattenToRawCallData,
  transformStringArrayToInteger,
} from './types/helper';
import { CallbackReturnType } from './ABIForm';
import { formsAtom } from './atoms';

const typeToTagColor = (name: string): TagColors => {
  try {
    const pathNames = name?.split('::');
    const coreType = pathNames[pathNames.length - 1];
    // console.log({pathNames, coreType})

    switch (coreType) {
      case 'u8':
        return 'green';
      case 'u16':
        return 'green';
      case 'u32':
        return 'green';
      case 'u64':
        return 'green';
      case 'u128':
        return 'green';
      case 'bool':
        return 'indigo';
      case 'felt252':
        return 'yellow';
      case 'ContractAddress':
        return 'pink';
      default:
        return 'blue';
    }
  } catch (e) {
    return 'blue';
  }
};

type IParseInputFieldsFromObject = {
  abiTypes: Record<string, string | {} | Array<{}>>;
  errors: Record<string, string | {} | Array<{}>>;
  handleArrayPop: (path: string[], index: number) => void;
  handleArrayPush: (path: string[], value: string | {}) => void;
  handleChange: (e: React.ChangeEvent<any>) => any;
  initialValues: Record<string, string | {} | Array<{}>>;
  parentKeys?: string[];
  values: Record<string, string | {} | Array<{}>>;
};
const ParseInputFieldsFromObject: React.FC<IParseInputFieldsFromObject> = ({
  values,
  errors,
  abiTypes,
  initialValues,
  parentKeys,
  handleChange,
  handleArrayPush,
  handleArrayPop,
}) => {
  if (typeof values === 'object') {
    const keys = Object.keys(values);

    return keys.map((key) => {
      const currentValueObject = values[key];

      if (typeof currentValueObject === 'string') {
        let name =
          parentKeys && parentKeys?.length > 0
            ? parentKeys?.reduce((p, c) => {
                if (Number.isNaN(parseInt(c, 10))) {
                  return `${p}.${c}`;
                }
                return `${p}[${c}]`;
              }, '')
            : '';
        if (name.length > 0 && name.startsWith('.')) {
          name = name.slice(1);
        }

        const fullPath = parentKeys ? [...parentKeys, key] : [key];
        let error = '';

        let abiTypeInfo = '';
        if (loadashFp.has(errors, fullPath)) {
          error = loadashFp.get(errors, fullPath);
        }

        // Since Type conform to initial state,
        // For arrays index 0 will have the valid type.
        const initialFullPath = fullPath.map((pathItem) => {
          if (Number.isNaN(parseInt(pathItem, 10))) {
            return pathItem;
          }
          return '0';
        });
        if (loadashFp.has(abiTypes, initialFullPath)) {
          abiTypeInfo = loadashFp.get(abiTypes, initialFullPath);
        }

        return (
          <div
            className="my-2 w-full px-2 py-1 border-gray-200 border-2 rounded function-form-input-wrapper"
            key={fullPath?.join('|')}
          >
            <label
              htmlFor={`${name ? `${name}.` : ''}${key}`}
              className="block mb-2 text-sm font-medium input-label"
            >
              {`${name ? `${name}.` : ''}${key}`}
              <Tag
                style={{ marginLeft: '1rem' }}
                tag={typeToTagColor(abiTypeInfo)}
                className="input-tag"
              >
                {abiTypeInfo}
              </Tag>
            </label>
            <input
              type="text"
              id={`${name ? `${name}.` : ''}${key}`}
              name={`${name ? `${name}.` : ''}${key}`}
              placeholder={`${name ? `${name}.` : ''}${key}`}
              value={currentValueObject}
              className="w-full bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block p-2.5 function-form-input"
              onChange={handleChange}
            />
            <p className="input-error">{error}</p>
          </div>
        );
      }
      if (
        typeof currentValueObject === 'object' &&
        !Array.isArray(currentValueObject)
      ) {
        const lParentKeys = parentKeys ? [...parentKeys, key] : [key];
        return (
          <div
            className="w-full flex flex-col shadow-md shadow-green-500 rounded p-2 my-2 bg-green-50 function-struct"
            key={lParentKeys.join('|')}
          >
            <p className="text-xl font-bold function-struct-header">
              struct: {key}
            </p>
            <ParseInputFieldsFromObject
              values={{ ...currentValueObject }}
              errors={errors}
              initialValues={initialValues}
              abiTypes={abiTypes}
              parentKeys={lParentKeys}
              handleChange={handleChange}
              handleArrayPush={handleArrayPush}
              handleArrayPop={handleArrayPop}
            />
          </div>
        );
      }
      if (Array.isArray(currentValueObject)) {
        const pathKeys = parentKeys ? [...parentKeys, key] : [key];
        const initalArrPath = pathKeys.map((pathItem) => {
          if (Number.isNaN(parseInt(pathItem, 10))) {
            return pathItem;
          }
          return '0';
        });
        let initalObj = '';
        if (loadashFp.has(initialValues, initalArrPath)) {
          const initalArr = loadashFp.get(initialValues, initalArrPath);
          if (initalArr && Array.isArray(initalArr) && initalArr.length > 0) {
            [initalObj] = initalArr;
          }
        }

        const [accordianTabsState, setAccordianTabsState] = useState<string[]>(
          (): string[] => {
            const retArr = currentValueObject
              ? currentValueObject?.map((obj, index) => {
                  const lParentKeys = parentKeys
                    ? [...parentKeys, key, index.toString()]
                    : [key, index.toString()];
                  return lParentKeys.join('|');
                })
              : [];
            return retArr;
          }
        );

        return (
          <AccordionRoot
            type="multiple"
            key={`accordion-root|${pathKeys.join('|')}`}
            className="w-full shadow-sm shadow-purple-500 p-2 rounded bg-purple-50 function-array-root"
            value={accordianTabsState}
            onValueChange={(value) => {
              const diff = loadashFp.difference(accordianTabsState, value);
              if (diff.length > 0) {
                const filteredAccordianTabsState = accordianTabsState.filter(
                  (activeTab) => !diff.includes(activeTab)
                );
                setAccordianTabsState([...filteredAccordianTabsState]);
                return;
              }
              setAccordianTabsState([...accordianTabsState, ...value]);
            }}
          >
            <div className="flex justify-between items-center function-array-header">
              <h5 className="text-xl array-title">Array: {key}</h5>
              <Button
                onClick={(e) => {
                  e.stopPropagation();
                  handleArrayPush(pathKeys, initalObj);
                }}
                className="array-add"
              >
                ADD +
              </Button>
            </div>
            {currentValueObject?.map((obj, index) => {
              const lParentKeys = parentKeys
                ? [...parentKeys, key, index.toString()]
                : [key, index.toString()];

              if (typeof obj === 'string') {
                // In this case it is an array of coreDataType, so don't have to call recursively here.
                // Stop Condition

                let name =
                  lParentKeys && lParentKeys?.length > 0
                    ? lParentKeys?.reduce((p, c) => {
                        if (Number.isNaN(parseInt(c, 10))) {
                          return `${p}.${c}`;
                        }
                        return `${p}[${c}]`;
                      }, '')
                    : '';
                if (name.length > 0 && name.startsWith('.')) {
                  name = name.slice(1);
                }

                const fullPath = lParentKeys;

                let error = '';
                let coreArrAbiTypeInfo = '';

                if (loadashFp.has(errors, fullPath)) {
                  error = loadashFp.get(errors, fullPath);
                }

                // Since Type conform to initial state,
                // For arrays index 0 will have the valid type.
                const initialFullPathCoreArr = fullPath.map((pathItem) => {
                  if (Number.isNaN(parseInt(pathItem, 10))) {
                    return pathItem;
                  }
                  return '0';
                });
                if (loadashFp.has(abiTypes, initialFullPathCoreArr)) {
                  coreArrAbiTypeInfo = loadashFp.get(
                    abiTypes,
                    initialFullPathCoreArr
                  );
                }

                return (
                  <div
                    className="w-full flex flex-col items-end shadow-md shadow-green-500 rounded p-2 bg-green-50 my-2 array-core-item"
                    key={fullPath.join('|')}
                  >
                    <Button
                      className="w-max array-core-item-delete"
                      color="red"
                      onClick={(e) => {
                        e.stopPropagation();
                        handleArrayPop(pathKeys, index);
                      }}
                    >
                      DELETE
                    </Button>
                    <div className="my-2 w-full px-2 py-1 border-gray-200 border-2 rounded function-form-input-wrapper">
                      <label
                        htmlFor={`${name}`}
                        className="block mb-2 text-sm font-medium input-label"
                      >
                        {`${name}`}{' '}
                        <Tag
                          style={{ marginLeft: '1rem' }}
                          tag={typeToTagColor(coreArrAbiTypeInfo)}
                          className="input-tag"
                        >
                          {coreArrAbiTypeInfo}
                        </Tag>{' '}
                      </label>
                      <input
                        type="text"
                        name={`${name}`}
                        id={`${name}`}
                        value={obj}
                        placeholder={`${name}`}
                        className="w-full bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block p-2.5 function-form-input"
                        onChange={handleChange}
                      />
                      <p className="input-error">{error}</p>
                    </div>
                  </div>
                );
              }

              return (
                <AccordionItem
                  value={lParentKeys.join('|')}
                  className="w-full flex flex-col shadow-md shadow-green-500 rounded p-2 bg-green-50 my-2 array-complex-item"
                  key={lParentKeys.join('|')}
                >
                  <AccordionTrigger className="w-full hover:shadow-md hover:bg-slate-50 rounded array-complex-item-trigger">
                    <div className="flex justify-between items-center w-full px-2">
                      <p className="text-xl font-bold array-complex-item-header">
                        {index + 1}. struct: {key}
                      </p>
                      <div
                        className={`${ButtonColorsClasses.red} array-complex-item-delete`}
                        role="button"
                        tabIndex={0}
                        onClick={(e) => {
                          e.stopPropagation();
                          handleArrayPop(pathKeys, index);
                        }}
                        onKeyDown={(e) => {
                          e.stopPropagation();
                          handleArrayPop(pathKeys, index);
                        }}
                      >
                        DELETE -
                      </div>
                    </div>
                  </AccordionTrigger>
                  <AccordionContent className="array-complex-item-content">
                    <ParseInputFieldsFromObject
                      values={{ ...obj }}
                      errors={errors}
                      initialValues={initialValues}
                      abiTypes={abiTypes}
                      parentKeys={lParentKeys}
                      handleChange={handleChange}
                      handleArrayPush={handleArrayPush}
                      handleArrayPop={handleArrayPop}
                    />
                  </AccordionContent>
                </AccordionItem>
              );
            })}
          </AccordionRoot>
        );
      }
      return '';
    });
  }
  return <p className="failed-parse-type">Could not parse type!!</p>;
};

type IFunctionForm = {
  callbackFn: (value: CallbackReturnType) => void;
  functionAbi: ABIFunction;
  response?: React.ReactNode;
  structs: ABIStruct[];
  // enums: ABIEnum[];
};

const FunctionForm: React.FC<IFunctionForm> = ({
  functionAbi,
  structs,
  callbackFn,
  response,
  // enums,
}) => {
  // Check if functionAbi is correct with yup validation schema
  try {
    yupAbiFunctionSchema.validateSync(functionAbi);
  } catch (e) {
    console.error(e);
    return <p className="invalid-abi">Not a valid function ABI</p>;
  }

  const initialValuesMap = reduceFunctionInputs(functionAbi?.inputs, structs);
  // console.log({initialValuesMap, functionAbi, structs});
  // Freezing object, as these are reference maps used to make initial forms
  // also helpers like type info + validation schema
  const initialValues = Object.freeze(extractInitialValues(initialValuesMap));
  const validationSchema = Object.freeze(
    extractValidationSchema(initialValuesMap)
  );
  const abiTypesInfo = Object.freeze(extractAbiTypes(initialValuesMap));
  // console.log({ initialValues, validationSchema, abiTypesInfo });

  // Persistent State on Jotai
  const [formStates, setFormsState] = useAtom(formsAtom);
  const oldFormStates = formStates[functionAbi.name];

  // console.log(functionAbi?.name, { oldFormStates });

  const { values, errors, setValues, handleChange, handleSubmit } = useFormik({
    initialValues: {
      ...initialValues,
    },
    validationSchema: Yup.object(validationSchema),
    onSubmit: (finalValues) => {
      try {
        const rawArrayValues = flattenToRawCallData(finalValues);
        const starkliValues = transformStringArrayToInteger(
          flattenArrays(rawArrayValues) as string[]
        );

        const starknetValues = Object.keys(finalValues).map(
          // @ts-ignore
          (key) => finalValues[key]
        );

        const callbackReturnValues: CallbackReturnType = {
          raw: finalValues,
          functionName: functionAbi?.name,
          stateMutability: functionAbi?.state_mutability,
          starknetjs: starknetValues,
          starkli: {
            bigint: starkliValues,
            decimal: starkliValues.map((v) => v.toString(10)).join(' '),
            hex: starkliValues.map((v) => `0x${v.toString(16)}`).join(' '),
          },
        };
        callbackFn(callbackReturnValues);
      } catch (e) {
        console.error(e);
        callbackFn({
          raw: finalValues,
          functionName: functionAbi?.name,
          stateMutability: functionAbi?.state_mutability,
        });
      }
    },
  });

  useEffect(() => {
    if (oldFormStates) {
      setValues(oldFormStates);
    }
  }, []);

  useEffect(() => {
    // Update the Atom with a function name key
    if (!loadashFp.isEqual(initialValues, values)) {
      setFormsState({
        ...formStates,
        [functionAbi.name]: values,
      });
    }
  }, [values]);

  // console.log(values, errors, dirty);

  const handleArrayPush = (path: string[], value: string | {}) => {
    if (loadashFp.has(values, path)) {
      const oldValues = loadashFp.get(values, path);
      const newValues = [...oldValues, value];
      loadashFp.set(values, path, newValues);
      setValues({ ...values });
    }
  };

  const handleArrayPop = (path: string[], index: number) => {
    if (loadashFp.has(values, path)) {
      const oldValues = loadashFp.get(values, path);
      const newValues = [
        ...oldValues.filter((o: any, i: number) => i !== index),
      ];
      loadashFp.set(values, path, newValues);
      setValues({ ...values });
    }
  };

  return (
    <div className="bg-slate-100 p-3 rounded my-2 shadow-md function-root">
      <div className="flex items-center function-header">
        <p className="mr-2 text-md font-bold text-black  ">
          {functionAbi?.name}
        </p>
        (
        {functionAbi?.inputs?.map((ip) => (
          <React.Fragment key={ip?.name}>
            <span className="mr-2 text-sm font-normal dark:text-gray-400">
              {ip?.name}:
            </span>
            <Tag color={typeToTagColor(ip?.type)}>{ip?.type}</Tag>
          </React.Fragment>
        ))}
        )
      </div>
      <form onSubmit={handleSubmit} className="function-form">
        <ParseInputFieldsFromObject
          values={values}
          errors={errors}
          initialValues={initialValues}
          abiTypes={abiTypesInfo}
          handleChange={handleChange}
          handleArrayPush={handleArrayPush}
          handleArrayPop={handleArrayPop}
        />
        <Button
          type="submit"
          color="purple"
          className="my-2 function-form-submit"
        >
          Call
        </Button>
      </form>
      <div className="my-2 function-response">{response}</div>
    </div>
  );
};

export default FunctionForm;
