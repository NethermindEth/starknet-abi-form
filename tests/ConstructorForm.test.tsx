import React from 'react';
import renderer from 'react-test-renderer';
import { it, expect } from 'vitest';
import { CallbackReturnType, ConstructorForm } from '../src';

const sampleAbi = [
  {
    type: 'function',
    name: 'user_input',
    inputs: [
      {
        name: 'state',
        type: 'core::array::Array::<core::integer::u128>',
      },
    ],
    outputs: [],
    state_mutability: 'external',
  },
  {
    type: 'constructor',
    name: 'constructor',
    inputs: [
      {
        name: 'owner',
        type: 'core::felt252',
      },
      {
        name: 'guardian',
        type: 'core::felt252',
      },
    ],
  },
];

it('renders correctly', () => {
  const callback = (val: CallbackReturnType) => val;
  const tree = renderer
    .create(<ConstructorForm abi={sampleAbi} callBackFn={callback} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
