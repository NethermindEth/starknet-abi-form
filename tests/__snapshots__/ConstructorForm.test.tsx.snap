// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders correctly 1`] = `
<div
  className="bg-slate-100 p-3 rounded my-2 shadow-md function-root"
>
  <div
    className="flex items-center function-header"
  >
    <p
      className="mr-2 text-md font-bold text-black  "
    >
      constructor
    </p>
    (
    <span
      className="mr-2 text-sm font-normal dark:text-gray-400"
    >
      owner
      :
    </span>
    <span
      className="bg-blue-100 text-blue-800 text-xs font-medium mr-2 px-2.5 py-0.5 rounded dark:bg-blue-900 dark:text-blue-300 tag-blue undefined"
      color="yellow"
    >
      core::felt252
    </span>
    <span
      className="mr-2 text-sm font-normal dark:text-gray-400"
    >
      guardian
      :
    </span>
    <span
      className="bg-blue-100 text-blue-800 text-xs font-medium mr-2 px-2.5 py-0.5 rounded dark:bg-blue-900 dark:text-blue-300 tag-blue undefined"
      color="yellow"
    >
      core::felt252
    </span>
    )
  </div>
  <form
    className="function-form"
    onSubmit={[Function]}
  >
    <div
      className="my-2 w-full px-2 py-1 border-gray-200 border-2 rounded function-form-input-wrapper"
    >
      <label
        className="block mb-2 text-sm font-medium input-label"
        htmlFor="owner"
      >
        owner
        <span
          className="bg-yellow-100 text-yellow-800 text-xs font-medium mr-2 px-2.5 py-0.5 rounded dark:bg-yellow-900 dark:text-yellow-300 tag-yellow input-tag"
          style={
            {
              "marginLeft": "1rem",
            }
          }
        >
          core::felt252
        </span>
      </label>
      <input
        className="w-full bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block p-2.5 function-form-input"
        id="owner"
        name="owner"
        onChange={[Function]}
        placeholder="owner"
        type="text"
        value=""
      />
      <button
        data-state="closed"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onPointerDown={[Function]}
        onPointerLeave={[Function]}
        onPointerMove={[Function]}
      >
        <p
          className="tooltip-input-final-value"
        />
      </button>
      <p
        className="input-error"
      />
    </div>
    <div
      className="my-2 w-full px-2 py-1 border-gray-200 border-2 rounded function-form-input-wrapper"
    >
      <label
        className="block mb-2 text-sm font-medium input-label"
        htmlFor="guardian"
      >
        guardian
        <span
          className="bg-yellow-100 text-yellow-800 text-xs font-medium mr-2 px-2.5 py-0.5 rounded dark:bg-yellow-900 dark:text-yellow-300 tag-yellow input-tag"
          style={
            {
              "marginLeft": "1rem",
            }
          }
        >
          core::felt252
        </span>
      </label>
      <input
        className="w-full bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block p-2.5 function-form-input"
        id="guardian"
        name="guardian"
        onChange={[Function]}
        placeholder="guardian"
        type="text"
        value=""
      />
      <button
        data-state="closed"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onPointerDown={[Function]}
        onPointerLeave={[Function]}
        onPointerMove={[Function]}
      >
        <p
          className="tooltip-input-final-value"
        />
      </button>
      <p
        className="input-error"
      />
    </div>
    <button
      className="text-white bg-gradient-to-r from-purple-500 via-purple-600 to-purple-700 hover:bg-gradient-to-br focus:ring-4 focus:outline-none focus:ring-purple-300 dark:focus:ring-purple-800 font-medium rounded-lg text-sm px-5 py-2.5 text-center btn-purple my-2 function-form-submit"
      type="submit"
    >
      Call
    </button>
  </form>
  <div
    className="my-2 function-response"
  />
</div>
`;
